File: app/src/main/java/org/kbieron/iomerge/fragments/ServerListFragment.java
package org.kbieron.iomerge.fragments;

import android.app.Fragment;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;
import android.widget.TextView;
import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.ViewById;
import org.kbieron.iomerge.android.R;
import org.kbieron.iomerge.database.ServerBean;
import org.kbieron.iomerge.database.ServerDAO;
import org.kbieron.iomerge.services.NetworkManager;
import org.kbieron.iomerge.services.NetworkManager_;


@EFragment(R.layout.server_list_fragment_layout)
public class ServerListFragment extends Fragment implements OnServerListAdapter.ItemClickListener {

	@ViewById(R.id.server_list_view)
	RecyclerView recyclerView;

	@ViewById(R.id.fab)
	FloatingActionButton floatingActionButton;

	@ViewById(R.id.address)
	TextView addressView;

	@ViewById(R.id.port)
	TextView portView;

	@Bean
	ServerDAO serverDAO;
	private OnServerListAdapter listAdapter;


	@AfterViews
	public void createList() {

		// setup recycler view
		listAdapter = new OnServerListAdapter(serverDAO);
		listAdapter.setItemClickListener(this);
		recyclerView.setAdapter(listAdapter);
		recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
		new ItemTouchHelper(new SwipeCallback())
				.attachToRecyclerView(recyclerView);

		// setup fab
		floatingActionButton.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				AddServerDialogFragment addServerDialogFragment = new AddServerDialogFragment();
				addServerDialogFragment.show(getFragmentManager(), "add server");
			}
		});
		floatingActionButton.show();
	}

	@Override
	public void onItemClick(ServerBean server) {

		NetworkManager_.intent(getActivity())
				.action(NetworkManager.CONNECT_ACTION)
				.extra(NetworkManager.SERVER_EXTRA, server)
				.start();
	}

	private void removeServer(ServerBean server) {
		serverDAO.deleteServer(server);
		listAdapter.removeServer(server);
	}

	private class SwipeCallback extends ItemTouchHelper.SimpleCallback {

		SwipeCallback() {
			super(0, ItemTouchHelper.LEFT);
		}

		@Override
		public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
			return false;
		}

		@Override
		public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
			removeServer(((OnServerListAdapter.ServerViewHolder) viewHolder).getServer());
		}

	}
}

File: app/src/main/java/org/kbieron/iomerge/fragments/ServerListFragment.kt
package org.kbieron.iomerge.fragments

import android.app.Fragment
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.pawegio.kandroid.IntentFor
import kotterknife.bindView
import org.kbieron.iomerge.android.R
import org.kbieron.iomerge.database.ServerBean
import org.kbieron.iomerge.database.ServerDAO
import org.kbieron.iomerge.services.NetworkManager


open class ServerListFragment : Fragment(), OnServerListAdapter.ItemClickListener {

	private val recyclerView by bindView<RecyclerView>(R.id.server_list_view)

	private val floatingActionButton by bindView<FloatingActionButton>(R.id.fab)

	private val serverDAO by lazy { ServerDAO(activity) }

	private val listAdapter by lazy { OnServerListAdapter(serverDAO, this) }

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View
			= inflater.inflate(R.layout.server_list_fragment_layout, container, false)

	override fun onResume() {
		super.onResume()
		createList()
		setupFAB()
	}

	private fun createList() {
		recyclerView.apply {
			adapter = listAdapter
			layoutManager = LinearLayoutManager(activity)
		}
		ItemTouchHelper(SwipeCallback()).attachToRecyclerView(recyclerView)
	}

	private fun setupFAB() {
		floatingActionButton.apply {
			setOnClickListener {
				val addServerDialogFragment = AddServerDialogFragment()
				addServerDialogFragment.show(fragmentManager, "add server")
			}
		}.show()
	}

	override fun onItemClick(server: ServerBean) {
		IntentFor<NetworkManager>(activity).apply {
			action = NetworkManager.CONNECT_ACTION
			putExtra(NetworkManager.SERVER_EXTRA, server)

		}.let { activity.startService(it) }
	}

	private fun removeServer(server: ServerBean?) {
		server?.let {
			serverDAO.deleteServer(it)
			listAdapter.removeServer(it)
		}
	}

	private inner class SwipeCallback() : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {

		override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
			return false
		}

		override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
			removeServer((viewHolder as OnServerListAdapter.ServerViewHolder).server)
		}

	}
}